components/esp_dns/esp_dns.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * esp_dns.c - Custom DNS module for ESP32 with multiple protocol support
+ *
+ * Provides DNS resolution capabilities with support for various protocols:
+ * - Standard UDP/TCP DNS (Port 53)
+ * - DNS over TLS (DoT) (Port 853)
+ * - DNS over HTTPS (DoH) (Port 443)
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/semphr.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+//#include "esp_netif.h"
+#include "esp_system.h"
+#include "lwip/api.h"
+#include "lwip/dns.h"
+#include "lwip/sockets.h"
+#include "lwip/netdb.h"
+
+#include "esp_dns.h"
+#include "esp_dns_priv.h"
+#include "esp_dns_udp.h"
+#include "esp_dns_tcp.h"
+#include "esp_dns_dot.h"
+#include "esp_dns_doh.h"
+
+#define TAG "ESP_DNS"
+
+
+/* Global DNS handle instance */
+static esp_dns_handle_t g_dns_handle = NULL;
+
+/* Mutex for protecting global handle access */
+static SemaphoreHandle_t g_dns_global_mutex = NULL;
+
+
+esp_dns_handle_t dns_create_handle(void)
+{
+    static struct dns_handle instance;
+    static bool initialized = false;
+
+    if (!initialized) {
+        memset(&instance, 0, sizeof(instance));
+        initialized = true;
+    }
+
+    return &instance;
+}
+
+/**
+ * Initialize the DNS module with provided configuration.
+ *
+ * @param config Pointer to esp_dns_config_t structure with desired DNS configuration
+ * @return On success, returns a handle to the initialized DNS module.
+ *         On failure, returns NULL.
+ */
+esp_dns_handle_t esp_dns_init(const esp_dns_config_t *config)
+{
+    /* Validate parameters */
+    if (config == NULL) {
+        ESP_LOGE(TAG, "Invalid configuration (NULL)");
+        return NULL;
+    }
+
+    /* Create global mutex if it doesn't exist */
+    if (g_dns_global_mutex == NULL) {
+        g_dns_global_mutex = xSemaphoreCreateMutex();
+        if (g_dns_global_mutex == NULL) {
+            ESP_LOGE(TAG, "Failed to create global mutex");
+            return NULL;
+        }
+    }
+
+    /* Take the global mutex */
+    if (xSemaphoreTake(g_dns_global_mutex, portMAX_DELAY) != pdTRUE) {
+        ESP_LOGE(TAG, "Failed to take global mutex");
+        return NULL;
+    }
+
+    /* Check if we need to clean up an existing handle */
+    if (g_dns_handle != NULL) {
```

Comment:
You are setting a global handle and giving it back to the user. A better approach would be to force the user to clean up first and return an error here.

Reply:
Add error saying the cleanup is needed.

Reply:
Done